### ISSUE TITLE
armhf build?

### ISSUE THREAD

**Issue by webash** [Microsoft/OMS-Agent-for-Linux#20]
Any chance of an armhf build to be able to run this on embedded / small-scale linux environments, like the Raspberry Pi?

**Comment by jeffaco** [comment]
Note that OMS-Agent is dependent on Ruby, and I have no idea if Ruby runs on the Rasberry Pi.

Our primary dependencies are Ruby and Fluentd (Fluentd runs on top of Ruby), as well as a little bit of shell code (not much) and OpenSSL dependencies (for security).

If you can get Ruby and Fluentd going, I'll try and help you out the rest of the way!

**Comment by webash** [comment]
I have zero experience with ruby, but [this article](http://rayhightower.com/blog/2012/12/03/ruby-on-raspberry-pi/) seems to suggest its possible. If I get some time tonight, I'll see if I can get Ruby installed. Any suggestion for a hello-world-type script that would determine whether I have Ruby running sufficiently?

**Comment by andschwa** [comment]
Fluentd [supports](http://docs.fluentd.org/articles/raspberrypi-cloud-data-logger) Raspberry Pi devices (at least those running Raspbian). At that point, they're not much different from a normal Debian machine, which means it will have OpenSSL as well.

We actually used Fluentd on a Pi at OneWeek to aggregate sniffed packets of some archaic protocols.

**Comment by webash** [comment]
Great! I'll install Ruby and Fluentd tonight... what would I need to do from there to get OMS installed/running?

**Comment by andschwa** [comment]
That's a good question. @jeffaco do you know where OMS's build-from-source instructions are? I couldn't find them.

**Comment by jeffaco** [comment]
It turns out that you don't need to install Ruby as such, or Fluentd as such. We build it as part of the OMS build (we own the packages, so we build them and install in our directory). This means that the kits that we provide won't work on the Raspberry Pi, as our kits are based on the PC architecture.

You need to build from source. And, unfortunately, we haven't posted instructions on how to build from source because the bits we need aren't fully open source yet. Project [Build-OMS-Agent-for-Linux](https://github.com/Microsoft/Build-OMS-Agent-for-Linux) is actually a superproject to grab all of our dependencies so you can build, but a few of those are still private pending legal approval to open them up.

I've been asking @agup006 about this (he's talking to legal) every few days.

Anurag, what do the lawyers have to say? What do we tell folks that need to be able to build?

If worse comes to worse, you could actually install Ruby and Fluentd, and then take our [fluentd plugins](https://github.com/Microsoft/OMS-Agent-for-Linux/tree/master/source/code/plugins) and make minor edits (path and stuff) to make them work in that environment. But if you did that, you wouldn't have our machine stats. All you'd really be able to do is log things, I think (no cpu statistics, for example).

What, exactly, would you like to do with the Raspberry Pi? What sort of information do you want to log?

**Comment by webash** [comment]
I'm looking to provision Raspberry Pis as a 'micro-server' in low power/commodity environments, but I still need to ensure the device is online, performant and be able to collate its error messages to somewhere useful. OMS is the first service I turned to for this since I have a heavily Microsoft infrastructure background. Plus, this trial with Raspberry Pi should give me the insight I need to be able to recommend it to my clients at work; or possibly even use it for our Managed Services offering ourselves.

**Comment by jeffaco** [comment]
I'd like to better understand *exactly* what you're looking for from OMS-Agent on the Raspberry Pi.

You say that you need to ensure that the device:

- Is online
- Is Performant
- Is able to collate its error messages to somewhere useful

By "performant", you mean that do you do want statistical information (memory & cpu statistics) uploaded to OMS? Or were you after something else here?

As far as "is online", OMS won't give real-time information on that. But you could deduce issues in this area by lack of data being collected, but only if you're collecting data at some regular interval (like via performance statistics). Is this what you were looking for, or did you want something different?

The last part (able to collate its error messages), that's much easier, and you wouldn't need to build us from source for that. You could just install our plugins on top of fluentd itself (without installing our version of it).

Let me know the answers to my questions, and I can advise you further. Thanks.

**Comment by webash** [comment]
You're correct on the definition of performant - I need to have historical statistical data showing CPU and memory usage over time.

Deducing that an object is offline simply by a lack of data is enough for now, but ideally 'missed heartbeat' alerts would be useful.

The latter (error messages) is the most useful to me, so if there's a way I can proceed to achieving that first, then I'm all ears :ear:

**Comment by jeffaco** [comment]
For the error messages, what I would suggest is compiling both Ruby and fluentd for the Raspberry Pi, and getting both of those running on the platform. The fluentd project has some basic consistency tests that you can run to make sure that it's basically operating correctly. Once that is done, then you'll need to:

1. Copy our plugins to the fluentd plugin directory. These are available on [GitHub](https://github.com/Microsoft/OMS-Agent-for-Linux/tree/master/source/code/plugins),
2. Get a fluentd configuration file in place. Start with something similar to what runs on Linux on our OMS agent, but you'll need to change paths since you'll likely be running out of it's default directories,
3. Try running it. I'd expect to see errors, if you did everything properly, because you weren't on-boarded. If you get that far, you're doing great!
4. You'll need to manually onboard. This is done via [omsadmin.sh](https://github.com/Microsoft/OMS-Agent-for-Linux/blob/master/installer/scripts/omsadmin.sh). Edit that script due to changes in paths, then run it via the `-w <workspace id> -s <shared key>` qualifiers. You'll need to get the workspace_id and shared_key from the OMS website.

Note that what we provide streamlines all of this into a nice shell bundle that you can install. And we do provide other stuff (performance stats, etc). But for error logging, that's all you need.

I can answer specific questions if you get stuck somewhere and want to spend the time figuring it out. Your work will likely not be in vein, as you may need to deal with some of these issues anyway even if you compile our stuff from source.

@agup006, any progress on open source efforts?

**Comment by jeffaco** [comment]
Have we answered your questions? I'd like to close this issue if we have. You can reopen it later if further questions come up relating to this.

Please let us know, thanks!

**Comment by webash** [comment]
Unfortunately that's a bit more involved than I can be with this at the moment. But I'm hoping to get the agent rolled out on some Windows machines instead soon. I'll come back to the RPi issue specifically when I have some better time to thrash it out. Thanks for your help so far and let's hope you're able to Open Source the other bits that prevent full build/deploy on non AMD64 architectures. :)
